File: 8/python/main.py
class Solution:
    def myAtoi(self, s: str) -> int:
        ans = 0
        m = 1
        p = len(s) - 1
        start_flag = False
        for i, c in enumerate(s):
            is_number = ord("0") <= ord(c) <= ord("9")

            if not start_flag:
                if not (is_number or c in [" ", "+", "-"]):
                    break

                if c == " ":
                    continue

                if c == "-" or c == "+":
                    m = -1 if c == "-" else 1
                    start_flag = True
                    continue

                if is_number:
                    start_flag = True

            if start_flag:
                if is_number:
                    u = (ord(c) - ord("0")) * 10 ** (p - i)

                    ans += u

                if not is_number:
                    ans = ans // 10 ** (len(s) - i)
                    break

        if ans >= 2**31:
            return -(2**31) if m == -1 else 2**31 - 1
        return m * ans


s = Solution()

print(s.myAtoi(s="42"))
print(s.myAtoi(s=" -042"))
print(s.myAtoi(s="1337c0d3"))
print(s.myAtoi(s="0-1"))
print(s.myAtoi(s="+-12"))

File: 8/go/main.go
package main

import (
	"fmt"
	"math"
	"slices"
)

func myAtoI(s string) int {
	var m int64 = 1
	var ans int64 = 0
	flag := false

	// p := len(s) - 1

	for _, c := range s {
		isNum := '0' <= c && c <= '9'

		if !flag {
			if !(isNum || slices.Contains([]rune{'+', '-', ' '}, c)) {
				return 0
			}

			if c == ' ' {
				continue
			}

			if c == '-' || c == '+' {
				if c == '-' {
					m = -1
				}
				flag = true
				continue
			}

			if isNum {
				flag = true
			}
		}

		if flag {
			if isNum {
				ans *= 10
				ans += int64(c - '0')
			}

			if !isNum {
				break
			}
		}
	}

	if ans >= math.MaxInt32 {
		if m == -1 {
			return math.MinInt32
		}

		return math.MaxInt32
	}

	return int(m * ans)
}

func main() {
	// fmt.Println(myAtoI("42"))
	// fmt.Println(myAtoI(" -042"))
	// fmt.Println(myAtoI("1337c0d3"))
	// fmt.Println(myAtoI("0-1"))
	fmt.Println(myAtoI("9223372036854775808"))
}
